### PR TITLE
e2e tests for websockets proxy

### PR DESCRIPTION
**Pull request by alexeykazakov**

### CODE REVIEW

File: test/e2e/proxy_test.go [hunk 1]
```diff
@@ -188,3 +212,127 @@ func createPreexistingUserAndIdentity(t *testing.T, user proxyUsers) (*userv1.Us
...
+	user       proxyUser
+	connection *websocket.Conn
+	proxyHost  string
+
+	mu          sync.RWMutex
+	receivedCMs map[string]*corev1.ConfigMap
+}
+
+// start creates a new WebSocket connection. The method returns a function which is to be used to close the connection when done.
+func (w *wsWatcher) Start() func() {
```
Comment (xcoulon) [inline test/e2e/proxy_test.go:241]: this method doesn't need to be exported

File: test/e2e/proxy_test.go [hunk 2]
```diff
@@ -188,3 +212,127 @@ func createPreexistingUserAndIdentity(t *testing.T, user proxyUsers) (*userv1.Us
...
+		err = json.Unmarshal(msg, &message)
+		require.NoError(w.t, err)
+		copyCM := message.ConfigMap
+		w.mu.Lock()
+		w.receivedCMs[copyCM.Name] = &copyCM
+		w.mu.Unlock()
+	}
+}
+
+func (w *wsWatcher) WaitForConfigMap(retryInterval, timeout time.Duration, expectedCMName string) (*corev1.ConfigMap, error) {
```
Comment (xcoulon) [inline test/e2e/proxy_test.go:329]: this method doesn't need to be exported

File: test/e2e/proxy_test.go [hunk 3]
```diff
@@ -88,13 +98,27 @@ func TestProxyFlow(t *testing.T) {
...
+				defer closeConnection()
 
 				// when
 				proxyCl := hostAwait.CreateAPIProxyClient(user.token)
 				err := proxyCl.Create(context.TODO(), expectedCM)
 				require.NoError(t, err)
 
 				// then
+				// wait for the websocket watcher which uses the proxy to receive the config map
+				found, err := w.WaitForConfigMap(
```
Comment (xcoulon) [inline test/e2e/proxy_test.go:113]: what about testing that multiple resources can be received with the same websocket connection?

**Review by xcoulon — APPROVED** [review]

**Review by MatousJobanek — APPROVED** [review]
Nice :+1:

File: test/e2e/proxy_test.go (hunk 1 above)
Comment (alexeykazakov) [reply test/e2e/proxy_test.go:241]: That's right but I decided to export `Start()` and `WaitForConfigMap()` as the only "public" API of the watcher struct. Only those two methods are supposed to be used by clients. Other methods only to be used by the watcher itself.
So if you don't mind I would keep it as is in attempt to make it more clear ("external" vs. internal API).

File: test/e2e/proxy_test.go (hunk 2 above)
Comment (alexeykazakov) [reply test/e2e/proxy_test.go:329]: See above.

**Comment by alexeykazakov** [comment]
@xcoulon @MatousJobanek thank you for the review!

**Pull request merged**